File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/domain/usecase/GetRecipeDetailsFromFormUseCaseTest.java
package com.tarbadev.witchcraft.domain.usecase;

import com.tarbadev.witchcraft.domain.entity.Ingredient;
import com.tarbadev.witchcraft.domain.entity.Recipe;
import com.tarbadev.witchcraft.domain.entity.Step;
import com.tarbadev.witchcraft.domain.usecase.ConvertAndAddSameIngredientUseCase;
import com.tarbadev.witchcraft.domain.usecase.GetRecipeDetailsFromFormUseCase;
import com.tarbadev.witchcraft.domain.usecase.IngredientFromStringUseCase;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;

@RunWith(SpringRunner.class)
@SpringBootTest
@ActiveProfiles("test")
public class GetRecipeDetailsFromFormUseCaseTest {
  @Autowired private IngredientFromStringUseCase ingredientFromStringUseCase;
  @Autowired private ConvertAndAddSameIngredientUseCase convertAndAddSameIngredientsUseCase;

  private GetRecipeDetailsFromFormUseCase subject;

  @Before
  public void setUp() {
    subject = new GetRecipeDetailsFromFormUseCase(ingredientFromStringUseCase, convertAndAddSameIngredientsUseCase);
    Mockito.reset(convertAndAddSameIngredientsUseCase);
  }

  @Test
  public void execute() {
    String name = "Some recipe name";
    String originUrl = "http://some/originUrl/of/recipe";
    String imgUrl = "http://some/originUrl/of/recipe.png";
    String ingredients = String.join("\n"
        , "10 tbsp sugar"
        , "1/2 cup olive oil"
        , "1 lemon"
    );
    String steps = String.join("\n"
        , "Add ingredients and stir"
        , "Serve on each plate"
    );

    Recipe recipe = Recipe.builder()
        .name(name)
        .originUrl(originUrl)
        .imgUrl(imgUrl)
        .ingredients(Arrays.asList(
            Ingredient.builder()
                .name("sugar")
                .quantity(10.0)
                .unit("tbsp")
                .build(),
            Ingredient.builder()
                .name("olive oil")
                .quantity(0.5)
                .unit("cup")
                .build(),
            Ingredient.builder()
                .name("lemon")
                .quantity(1.0)
                .unit("")
                .build()
        ))
        .steps(Arrays.stream(steps.split("\n"))
            .map(step -> Step.builder().name(step).build())
            .collect(Collectors.toList()))
        .build();

    given(convertAndAddSameIngredientsUseCase.execute(recipe.getIngredients()))
        .willReturn(recipe.getIngredients());

    assertThat(subject.execute(name, originUrl, ingredients, steps, imgUrl)).isEqualTo(recipe);
  }

  @Test
  public void execute_addsSameIngredients() {
    String name = "Some recipe name";
    String url = "http://some/originUrl/of/recipe";
    String imgUrl = "http://some/originUrl/of/recipe.png";
    String ingredients = String.join("\n"
        , "10 tbsp sugar"
        , "1/2 cup olive oil"
        , "3 tsp olive oil"
        , "1 lemon"
    );
    String steps = String.join("\n"
        , "Add ingredients and stir"
        , "Serve on each plate"
    );

    List<Ingredient> allIngredients = Arrays.asList(
        Ingredient.builder()
            .name("sugar")
            .quantity(10.0)
            .unit("tbsp")
            .build(),
        Ingredient.builder()
            .name("olive oil")
            .quantity(0.5)
            .unit("cup")
            .build(),
        Ingredient.builder()
            .name("olive oil")
            .quantity(3.0)
            .unit("tsp")
            .build(),
        Ingredient.builder()
            .name("lemon")
            .quantity(1.0)
            .unit("")
            .build()
    );

    List<Ingredient> expectedIngredients = Arrays.asList(
        Ingredient.builder()
            .name("sugar")
            .quantity(10.0)
            .unit("tbsp")
            .build(),
        Ingredient.builder()
            .name("olive oil")
            .quantity(0.5625)
            .unit("cup")
            .build(),
        Ingredient.builder()
            .name("lemon")
            .quantity(1.0)
            .unit("")
            .build()
    );
    Recipe recipe = Recipe.builder()
        .name(name)
        .originUrl(url)
        .imgUrl(imgUrl)
        .ingredients(
            expectedIngredients)
        .steps(Arrays.stream(steps.split("\n"))
            .map(step -> Step.builder().name(step).build())
            .collect(Collectors.toList()))
        .build();

    given(convertAndAddSameIngredientsUseCase.execute(allIngredients)).willReturn(expectedIngredients);

    assertThat(subject.execute(name, url, ingredients, steps, imgUrl)).isEqualTo(recipe);
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/recipes/domain/usecase/GetRecipeDetailsFromFormUseCaseTest.kt
package com.tarbadev.witchcraft.recipes.domain.usecase

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.reset
import com.nhaarman.mockitokotlin2.whenever
import com.tarbadev.witchcraft.converter.cup
import com.tarbadev.witchcraft.converter.tablespoon
import com.tarbadev.witchcraft.converter.teaspoon
import com.tarbadev.witchcraft.converter.unit
import com.tarbadev.witchcraft.recipes.domain.entity.Ingredient
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.recipes.domain.entity.Step
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class GetRecipeDetailsFromFormUseCaseTest {
  private val ingredientFromStringUseCase: IngredientFromStringUseCase = mock()
  private val convertAndAddSameIngredientsUseCase: ConvertAndAddSameIngredientUseCase = mock()

  private lateinit var subject: GetRecipeDetailsFromFormUseCase

  @BeforeEach
  fun setUp() {
    subject = GetRecipeDetailsFromFormUseCase(ingredientFromStringUseCase, convertAndAddSameIngredientsUseCase)
    reset(
        convertAndAddSameIngredientsUseCase,
        ingredientFromStringUseCase
    )
  }

  @Test
  fun execute() {
    val name = "Some recipe name"
    val originUrl = "http://some/originUrl/of/recipe"
    val imgUrl = "http://some/originUrl/of/recipe.png"
    val ingredients = arrayOf("10 tbsp sugar", "1/2 cup olive oil", "1 lemon").joinToString("\n")
    val steps = arrayOf("Add ingredients and stir", "Serve on each plate").joinToString("\n")
    val portions = 4

    val ingredient1 = Ingredient(
        name = "sugar",
        quantity = 10.tablespoon
    )
    val ingredient2 = Ingredient(
        name = "olive oil",
        quantity = 0.5.cup
    )
    val ingredient3 = Ingredient(
        name = "lemon",
        quantity = 1.unit
    )
    val recipe = Recipe(
        name = name,
        originUrl = originUrl,
        imgUrl = imgUrl,
        ingredients = listOf(
            ingredient1,
            ingredient2,
            ingredient3
        ),
        steps = steps.split("\n")
            .dropLastWhile { it.isEmpty() }
            .map { step -> Step(name = step) },
        portions = portions
    )


    whenever(ingredientFromStringUseCase.execute("10 tbsp sugar")).thenReturn(ingredient1)
    whenever(ingredientFromStringUseCase.execute("1/2 cup olive oil")).thenReturn(ingredient2)
    whenever(ingredientFromStringUseCase.execute("1 lemon")).thenReturn(ingredient3)
    whenever(convertAndAddSameIngredientsUseCase.execute(recipe.ingredients))
        .thenReturn(recipe.ingredients)

    assertEquals(recipe, subject.execute(name, originUrl, ingredients, steps, imgUrl, portions))
  }

  @Test
  fun execute_addsSameIngredients() {
    val name = "Some recipe name"
    val url = "http://some/originUrl/of/recipe"
    val imgUrl = "http://some/originUrl/of/recipe.png"
    val ingredients = arrayOf("10 tbsp sugar", "1/2 cup olive oil", "3 tsp olive oil", "1 lemon").joinToString("\n")
    val steps = arrayOf("Add ingredients and stir", "Serve on each plate").joinToString("\n")
    val portions = 4

    val ingredient1 = Ingredient(
        name = "sugar",
        quantity = 10.tablespoon
    )
    val ingredient2 = Ingredient(
        name = "olive oil",
        quantity = 0.5.cup
    )
    val ingredient3 = Ingredient(
        name = "olive oil",
        quantity = 3.teaspoon
    )
    val ingredient4 = Ingredient(
        name = "lemon",
        quantity = 1.unit
    )
    val allIngredients = listOf(
        ingredient1,
        ingredient2,
        ingredient3,
        ingredient4
    )

    val expectedIngredients = listOf(
        Ingredient(
            name = "sugar",
            quantity = 10.tablespoon
        ),
        Ingredient(
            name = "olive oil",
            quantity = 0.5625.cup
        ),
        Ingredient(
            name = "lemon",
            quantity = 1.unit
        )
    )
    val recipe = Recipe(
        name = name,
        originUrl = url,
        imgUrl = imgUrl,
        ingredients = expectedIngredients,
        steps = steps.split("\n")
            .dropLastWhile { it.isEmpty() }
            .map { step -> Step(name = step) },
        portions = portions
    )

    whenever(ingredientFromStringUseCase.execute("10 tbsp sugar")).thenReturn(ingredient1)
    whenever(ingredientFromStringUseCase.execute("1/2 cup olive oil")).thenReturn(ingredient2)
    whenever(ingredientFromStringUseCase.execute("3 tsp olive oil")).thenReturn(ingredient3)
    whenever(ingredientFromStringUseCase.execute("1 lemon")).thenReturn(ingredient4)
    whenever(convertAndAddSameIngredientsUseCase.execute(allIngredients)).thenReturn(expectedIngredients)

    assertEquals(recipe, subject.execute(name, url, ingredients, steps, imgUrl, portions))
  }
}